dev, v3.0: add large binlog operation FAQ (#1436)
pingcap/docs: dev/faq/data-migration.md

The `driver: bad connection` error indicates that anomalies have occurred in the connection between DM and the upstream TiDB database (such as network failure, TiDB restart and so on) and that the data of the current request has not yet been sent to TiDB at that moment.

When this type of error occurs in the current version, use `stop-task` to stop the task and then use `start-task` to restart the task. The automatic retry mechanism of DM will be improved later.

> ## What can I do when a replication task is interrupted with failing to get or parse binlog error `get binlog error ERROR 1236 (HY000), binlog checksum mismatch, data may be corrupted` etc. returned?

Reviewer: ```suggestion
## What can I do when a replication task is interrupted with failing to get or parse binlog errors like `get binlog error ERROR 1236 (HY000)`, `binlog checksum mismatch, data may be corrupted` returned?
```

> This error may occur during the incremental replication process of the DM if the binlog file in the upstream exceeds 4 GB, and the DM encounters a replication interruption in the processing of the file (including interruption caused by abnormalities in common pause or stop tasks). It is because that the DM needs to store the replicated binlog position. However, MySQL officially uses unit32 to store it, so the offset value of binlog position of the part exceeding 4 GB overflows and a wrong binlog position is stored. After the task or DM-worker is restarted, this wrong binlog position is used to re-parse the binlog or relay log. In this case, you need to follow these steps to manually recover replication:

Reviewer: ```suggestion
During the DM incremental replication process, this error might occur if the binlog file in the upstream exceeds 4 GB and DM encounters a replication interruption when it processes this binlog file (including ordinary pause or stop tasks and abnormal interruptions). This is because DM needs to store the replicated binlog position. However, MySQL officially uses uint32 to store it, so the offset values of binlog positions that exceed 4 GB overflow and DM stores a wrong binlog position. After the task or DM-worker is restarted, this wrong binlog position is used to re-parse the binlog or relay log. Then the above error occurs. In this case, perform the following steps to manually recover replication:
```

Reviewer: - may -> might (Because "may" has multiple meanings.)
- unit32 -> u**in**t32
- Avoid unnecessary "the"
- Try to make sentences simple and concise

> - First, determine whether the error occurs in the relay log write or the binlog replication/syncer unit replication(by the component information in the log error message). If the error occurs in the relay log module and the breakpoints saved by the binlog replication/syncer unit are correct, you can first stop the task and the DM-worker, then set the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log writes, the replication will automatically continue from the breakpoint after restarting the task.

Reviewer 2: ```suggestion
- First, determine whether the error occurs in the relay log write or the binlog replication/syncer unit replication(by the component information in the log error message). If the error occurs in the relay log module and the checkpoints saved by the binlog replication/syncer unit are correct, you can first stop the task and the DM-worker, then set the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log writes, the replication will automatically continue from the checkpoint after restarting the task.
```

Reviewer: @juliezhang1112 [Ref link from docs-cn](https://pingcap.com/docs-cn/v3.0/reference/tools/tidb-lightning/checkpoints/) like that in TiDB Lightning

Author: Thanks for the reference! @lilin90

> - First, determine whether the error occurs in the relay log write or the binlog replication/syncer unit replication(by the component information in the log error message). If the error occurs in the relay log module and the breakpoints saved by the binlog replication/syncer unit are correct, you can first stop the task and the DM-worker, then set the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log writes, the replication will automatically continue from the breakpoint after restarting the task.

Reviewer: ```suggestion
1. Determine whether the error occurs in the relay log write or in the binlog replication/syncer unit replication (by the component information in the log error message).

2. Take the corresponding step according to the place where the error occurs.

    - If the error occurs in the relay log module and the checkpoints saved by the binlog replication/syncer unit are correct, you can first stop the task and the DM-worker, then manually adjust the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log writes well, the replication automatically continues from the checkpoint after restarting the task.
```

> - If the relay log writes, and the next file is been written, the error occurs in an invalid binlog position when the binlog replication/syncer unit reads a relay log file exceeding 4 GB. At this time, you can stop the task and set the binlog position to a valid one of the relay log, such as `4`. Note that you need to adjust both the global checkpoint and the binlog position of each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. It works as soon as the larger than 4 GB file is replicated.

Reviewer 2: ```suggestion
- If the relay log writes without error and has been rotated to the next file, but the error occurs in an invalid binlog position when the binlog replication/syncer unit reads a relay log file exceeding 4 GB. At this time, you can stop the task and set the binlog position to a valid one of the relay log, such as `4`. Note that you need to adjust both the global checkpoint and the binlog position of each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the replication status. It works as soon as the larger than 4 GB file is replicated.
```

Reviewer: @amyangfei I made a small change in my suggestions.

> - If the relay log writes, and the next file is been written, the error occurs in an invalid binlog position when the binlog replication/syncer unit reads a relay log file exceeding 4 GB. At this time, you can stop the task and set the binlog position to a valid one of the relay log, such as `4`. Note that you need to adjust both the global checkpoint and the binlog position of each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. It works as soon as the larger than 4 GB file is replicated.

Reviewer 3: ```suggestion
    - If one relay log is written without error and the next log is being written, the error occurs in an invalid binlog position when the `binlog replication` or Syncer unit reads a relay log file exceeding `4 GB`. In this situation, you can stop the task and manually set the binlog position of the relay log to a valid one, such as to `4`. Note that you also need to adjust both the global checkpoint and the binlog position of each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. The task resumes after the oversized (over `4 GB`) file is replicated.
```

> - If the relay log writes, and the next file is been written, the error occurs in an invalid binlog position when the binlog replication/syncer unit reads a relay log file exceeding 4 GB. At this time, you can stop the task and set the binlog position to a valid one of the relay log, such as `4`. Note that you need to adjust both the global checkpoint and the binlog position of each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. It works as soon as the larger than 4 GB file is replicated.

Reviewer: ```suggestion
    - If the relay log writes well and has been rotated to the next file, the error occurs in an invalid binlog position when the binlog replication/syncer unit reads a relay log file exceeding 4 GB. At this time, you can stop the task and manually adjust the binlog position to a valid one of the relay log, such as `4`. Note that you need to adjust the binlog position of both the global checkpoint and each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. It works well after the larger than 4 GB file is replicated.
```

Reviewer: **Misunderstanding** of the Chinese version: **Note that you need to adjust both the global checkpoint and the binlog position of each table checkpoint.**

When this type of error occurs in the current version, use `stop-task` to stop the task and then use `start-task` to restart the task. The automatic retry mechanism of DM will be improved later.

## What can I do when a replication task is interrupted with failing to get or parse binlog errors like `get binlog error ERROR 1236 (HY000)` and `binlog checksum mismatch, data may be corrupted` returned?

During the DM incremental replication process, this error might occur if the binlog file in the upstream exceeds **4 GB**, and DM encounters a replication interruption when processing this binlog file (including interruption caused by anomalies in ordinary pause or stop tasks).

**Cause:** DM needs to store the replicated binlog position, and MySQL officially uses `uint32` to store it, so the binlog position of the file with offset exceeding **4 GB** overflows and an incorrect binlog position is stored. After the task or DM-worker is restarted, this incorrect binlog position is used to re-parse the binlog or relay log.

> In this case, you can consider the following solution to manually recover replication:

Reviewer: ```suggestion
In this case, manually recover replication using the following solution:
```

1. Determine whether the error occurs during the write of the relay log or replication of Binlog replication/Syncer unit (according to the component information in the log error message).

2. Take the following step according to the place where the error occurs.

>     - If the error occurs in the relay log module and the checkpoints saved by the Binlog replication/Syncer unit are correct, you can first stop the task and the DM-worker, then manually adjust the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log is written without error, the replication automatically resumes from the breakpoint after the task is restarted.

Reviewer: ```suggestion

    - If the error occurs in the relay log module and the checkpoints saved by the Binlog replication/Syncer unit are correct, you can first stop the task and the DM-worker, then manually adjust the binlog position of the relay meta to `4`, and restart the DM-worker to re-pull the relay log. If the relay log is written without error, the replication automatically resumes from the breakpoint after the task is restarted.
```

>     - If the relay log is written well, and has been rotated to the next file, the error occurs in an invalid binlog position when the Binlog replication/Syncer unit reads a relay log file exceeding **4 GB**. In this situation, you can stop the task and manually set the binlog position of the relay log to a valid one such as `4`. Note that you also need to adjust the binlog position of both global checkpoints and each table checkpoint. Set the safe-mode of the task to `true` to ensure reentrant execution. Then, you can restart the replication task and observe the status. The task resumes after the oversized (larger than **4 GB**) file is replicated.

Reviewer: @amyangfei "global checkpoints" or "the global checkpoint", which one is accurate?

Author: Verified by @amyangfei , “the global checkpoint” is accurate here.
